x/mongo/driver/compression.go

```diff
@@ -41,6 +40,50 @@ func getZstdEncoder(l zstd.EncoderLevel) (*zstd.Encoder, error) {
 	return encoder, nil
 }
 
+var zlibEncoders = &sync.Map{}
+
+func getZlibEncoder(l int) (*zlibEncoder, error) {
+	if v, ok := zlibEncoders.Load(l); ok {
+		return v.(*zlibEncoder), nil
+	}
+	b := bytes.NewBuffer(nil)
+	w, err := zlib.NewWriterLevel(b, l)
+	if err != nil {
+		return nil, err
+	}
+	encoder := &zlibEncoder{l: new(sync.Mutex), w: w, b: b}
+	zlibEncoders.Store(l, encoder)
+
+	return encoder, nil
+}
+
+type zlibEncoder struct {
+	l sync.Locker
+	w *zlib.Writer
+	b *bytes.Buffer
+}
+
+func (e *zlibEncoder) Encode(dst, src []byte) ([]byte, error) {
+	e.l.Lock()
+	defer e.l.Unlock()
+
+	defer func() {
+		e.b.Reset()
+		e.w.Reset(e.b)
+	}()
```

Comment:
What are your thoughts on removing the defer functoin here and executing the "Reset" methods at the top of the calls to "Encode". This seems marginally safer.